lib/epochtalk_server/models/role_permission.ex

```diff
@@ -59,27 +59,76 @@ defmodule EpochtalkServer.Models.RolePermission do
     do: Repo.insert_all(RolePermission, roles_permissions)
 
   ## for admin api use, modifying permissions for a role
-  # no permissions to modify
-  # def modify_by_role(role, []), do: {:error, "No permissions to modify"}
-  # def modify_by_role(role, [%Permission{}|_] = permissions) do
-  #   # change role permission for each permission
-  #     # check default value
-  #     # check new value
-  #     # if new value is different, set modified true
-  #     # if new value is same, set modified false
-  #   # update roles table
-  # end
-  # def modify_by_role(role, %RolePermission{} = permission) do
-  #   # change role permission
-  #     # check default value
-  #     # check new value
-  #     # if new value is different, set modified true
-  #     # if new value is same, set modified false
-  #   # update roles table
-  # end
+  def modify_by_role(
+        %Role{
+          id: role_id,
+          permissions: new_permissions,
+          priority_restrictions: priority_restrictions
+        } = _new_role
+      ) do
+    # if a permission is false, they're not included in the permissions
+    # if they're all false, permissions can be empty
+    old_role_permissions =
+      from(rp in RolePermission,
+        where: rp.role_id == ^role_id
+      )
+      |> Repo.all()
+
+    new_permissions = new_permissions |> Iteraptor.to_flatmap()
+
+    # change a permission if it's different
+    new_role_permissions =
+      Enum.reduce(old_role_permissions, [], fn %{
+                                                 permission_path: permission_path,
+                                                 value: old_value
+                                               } = old_role_permission,
+                                               acc ->
+        # check new value for permission_path
+        # if value is not there, set it to false
+        new_value = new_permissions[permission_path] || false
+        # if new value is different
+        new_role_permission =
+          if old_value != new_value do
+            # set modified true
+            %{role_id: role_id, permission_path: permission_path, modified: true}
+            # if new value is same, set modified false
+          else
+            %{role_id: role_id, permission_path: permission_path, modified: false}
+          end
+
+        [new_role_permission | acc]
+      end)
+
+    # update role permissions for this role
+    upsert_modified(new_role_permissions)
+
+    # update role's permissions
+    permissions = RolePermission.permissions_map_by_role_id(role_id)
+    Role.set_permissions(role_id, permissions)
+
+    # update role's priority_restrictions
+    Role.set_priority_restrictions(role_id, priority_restrictions)
+
+    # return success
+    {:ok, :success}
+  end
+
+  # Used to update the modified field of a `RolePermission` in the database (should already exist)
+  defp upsert_modified([]), do: {:error, "Role permission list is empty"}
```

Comment:
This probably isn't necessary since this is a private function and you are not checking for the error anywhere

Reply:
got rid of it ✅